### ISSUE TITLE
Enable NamedSet when using devtools and combine

### ISSUE THREAD

**Issue by rosostolato** [pmndrs/zustand#216]
When using devtools with combine, combine's `set` is not of type `NamedSet` so typescript does not permit to give a third parameter to `set` function. Is there a way to work around it?

![image](https://user-images.githubusercontent.com/27731641/97222903-c95c9100-17ad-11eb-8a7b-6f743270dcaa.png)

maybe exporting interface from `middleware.ts:31`, I can force it to use.

**Comment by dai-shi** [comment]
Good catch. We need to extend types in `combine`. It's trivial but could be done. Would you like to work on it?

**Comment by rosostolato** [comment]
I will give a try, but I'm not so good in typescript. So if anyone could help me, please be free

**Comment by dai-shi** [comment]
It would require basic knowledge around generics. I could help if it's too hard. In any case, we want ts experts to improve types for middleware more broadly.

**Comment by dai-shi** [comment]
@rosostolato Any progress? I can work on it if you like.

**Comment by rosostolato** [comment]
I didn't start yet... please, feel free! Thanks!

**Comment by dai-shi** [comment]
Alright! I'll try it and ask your help for testing.

**Comment by dai-shi** [comment]
I think #222 doesn't do the job. https://github.com/pmndrs/zustand/pull/233#discussion_r518441842
Let's reopen this.

**Comment by barelyhuman** [comment]
@dai-shi naive change but I'd like to know where the change would break.
https://codesandbox.io/s/react-typescript-forked-j0f8w?file=/src/App.tsx:169-183

https://github.com/barelyhuman/zustand/pull/11

**Comment by dai-shi** [comment]
> where the change would break.

It might not break, but it's not correct typing if you use `combine` without `devtools`. What do you think?

I don't know how the typing with middleware chaining works in general. We have another issue #212.

**Comment by barelyhuman** [comment]
> if you use combine without devtools

Having it as an optional field won't help? But then dependent typing is a little more complex and I'll go through TS docs again in terms of optimally doing this. 

Either way, I'll pick this up and maybe even go through #212 if it can be handled in a similar way

**Comment by dai-shi** [comment]
> Having it as an optional field won't help? 

I mean, TS accepts the optional field, but JS ignores it. So, basically, something like this?
```ts
const fn = (x: number) => x * 2
const fn2: (x: number, s?: string) => number = fn
```

Although this specific issue is about devtools and combine, there can be any combinations with future middlewares. It just feels bad that combine knows something specific about devtools. Maybe, I'm too concerned about it. Could we combine `combine` and `immer`?

Thanks for picking it up.

I hope someone else has feedback too.

**Comment by barelyhuman** [comment]
@dai-shi  i was thinking of creating 2 types, then a type for the function that checks what the params signature is using `X extends FunctionType1 ? FunctionType1 : FunctionType2` 

though that becomes a little more hard to read and manage so I'm still in doubt if I should do that

**Comment by dai-shi** [comment]
Does that work? I'd like to see a solution, from my personal interest.

I agree that we should consider avoiding too technical, hard-to-maintain types.
For example, `PartialState` in zustand is one that is too complicated (wish we had a better solution.)

**Comment by barelyhuman** [comment]
@dai-shi the `Pick` logic could be reduced but I don't wanna break it for certain cases that it's already working well for, 
Yeah, I'll push a solution for the above recommendation as a separate PR, or I'll do it on my fork and you can check it
